src/OrchardCore.Modules/OrchardCore.OpenId/Configuration/OpenIdValidationConfiguration.cs

```diff
@@ -0,0 +1,264 @@
+                    {
+                        if (!Uri.TryCreate(issuer, UriKind.Absolute, out Uri uri))
+                        {
+                            throw new SecurityTokenInvalidIssuerException("The token issuer is not valid.");
+                        }
+
+                        var tenant = _runningShellTable.Match(uri.Authority, uri.AbsolutePath);
+                        if (tenant == null || !string.Equals(tenant.Name, settings.Tenant, StringComparison.Ordinal))
+                        {
+                            throw new SecurityTokenInvalidIssuerException("The token issuer is not valid.");
+                        }
+
+                        return issuer;
+                    };
+                }
+            }
+            // Otherwise, simply set the authority to allow the JWT handler to retrieve
+            // the endpoint URLs/signing keys from the remote server's discovery endpoint.
+            else
+            {
+                options.RequireHttpsMetadata = settings.Authority.StartsWith(Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase);
+                options.Audience = settings.Audience;
+                options.Authority = settings.Authority;
+            }
+        }
+
+        public void Configure(JwtBearerOptions options) => Debug.Fail("This infrastructure method shouldn't be called.");
+
+        public void Configure(string name, OAuthValidationOptions options)
+        {
+            // Ignore validation handler instances that don't correspond to the instance managed by the OpenID module.
+            if (!string.Equals(name, OAuthValidationDefaults.AuthenticationScheme, StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            var settings = GetValidationSettingsAsync().GetAwaiter().GetResult();
+            if (settings == null)
+            {
+                return;
+            }
+
+            // If the tokens are issued by an authorization server located in a separate tenant,
+            // create a tenant-specific data protection provider using the shared host provider.
+            if (!string.IsNullOrEmpty(settings.Tenant) &&
+                !string.Equals(settings.Tenant, _shellSettings.Name, StringComparison.Ordinal))
+            {
+                options.DataProtectionProvider = _hostDataProtectionProvider.CreateProtector(settings.Tenant);
+            }
+
+            // Don't allow the current tenant to choose the valid audiences, as this would
+            // otherwise allow it to introspect tokens meant to be used with another tenant.
+            options.Audiences.Add(OpenIdConstants.Prefixes.Tenant + _shellSettings.Name);
+        }
+
+        public void Configure(OAuthValidationOptions options) => Debug.Fail("This infrastructure method shouldn't be called.");
+
+        private IOpenIdServerService GetServerService(string tenant)
```

Comment:
We'll need to introduce a way to easily and safely "lease" a service from a separate tenant context. We could probably use the existing APIs, but I would prefer something like:

```csharp
// This method would throw an exception if the tenant context was not in a valid state.
// Note: service is of type ShellService<IOpenIdServerService>.
using (var service = context.LeaseService<IOpenIdServerService>())
{
    // You can safely use service.Instance as the shell context guarantees
    // the tenant won't be released until the ShellService is released.
}
```